### ISSUE TITLE
BUG: Overlay struggles if `geometry` column not named geometry. 

### ISSUE THREAD

**Issue by nickeubank** [geopandas/geopandas#306]
Working on following code for docs:

    os.chdir('/users/nick/github/geopandas/doc/')
	import geopandas as gpd
	world = gpd.GeoDataFrame().from_file('source/_example_data/naturalearth_lowres.shp')
	capitals = gpd.GeoDataFrame().from_file('source/_example_data/naturalearth_cities.shp')
	   
	   # For spatial join
	countries = world[['geometry', 'name']]

	   # Project
	countries = countries.to_crs('+init=epsg:3395')[countries.name!="Antarctica"]
	capitals = capitals.to_crs('+init=epsg:3395')

	capitals['buffers']= capitals.buffer(500000)
	capitals = capitals.set_geometry('buffers')
	capitals.plot()

	from geopandas.tools import overlay
	country_cores = overlay(countries, capitals, how='intersection')

And overlay fails with error 

---------------------------------------------------------------------------
	TypeError                                 Traceback (most recent call last)
	<ipython-input-7-8966ae43ad89> in <module>()
	----> 1 country_cores = overlay(countries, capitals, how='intersection')

	/Users/Nick/github/geopandas/geopandas/tools/overlay.py in overlay(df1, df2, how, use_sindex)
	     85     # Collect the interior and exterior rings
	     86     rings1 = _extract_rings(df1)
	---> 87     rings2 = _extract_rings(df2)
	     88     mls1 = MultiLineString(rings1)
	     89     mls2 = MultiLineString(rings2)

	/Users/Nick/github/geopandas/geopandas/tools/overlay.py in _extract_rings(df)
	     35 
	     36         if geom.type not in ['Polygon', 'MultiPolygon']:
	---> 37             raise TypeError(poly_msg)
	     38 
	     39         if hasattr(geom, 'geoms'):

	TypeError: overlay only takes GeoDataFrames with (multi)polygon geometries

Source of the problem is that in what `iterrows` returns, the `geometry` *attribute* is lost, so it's just a name reference. 

Also, further support for #294!

**Comment by jorisvandenbossche** [comment]
@nickeubank Good catch (but regarding #294, changing the name of the 'geometry' column will not change anything here I think, you just have the same problem with another name). The issue is the usage of `df.iterrows` loosing the extra information of GeoSeries/GeoDataFrame

**Comment by jorisvandenbossche** [comment]
BTW, `iterrows` is really slow, and we should try to avoid its usage (http://pandas.pydata.org/pandas-docs/stable/basics.html#iterrows)

**Comment by nickeubank** [comment]
@jorisvandenbossche Re: #294 The reason it's related is that because we used the same name for the attribute and as the default column name, the implemented code "accidentally" worked. If they had had separate names, the existing code would have thrown an error. Yes, the failure comes from iterrows, but the reason it made it into the master branch was that we had allowed `df.geometry` to become overloaded. 

Want to open issue on removing iterrows? I don't think it's something I'd be able to address soon, but we can tag for 0.3 (unless you're feeling ambitious? :)).
